File: backend/hooks/validators/russian_business.py
#!/usr/bin/env python3
"""
Russian Business Validation Module
Validates INN, KPP, OGRN, and other Russian business identifiers
"""
import re
import sys
from typing import Dict, List, Tuple, Optional
from pathlib import Path


class RussianBusinessValidator:
    """Russian business validation utilities"""

    @staticmethod
    def validate_inn(inn: str) -> Tuple[bool, str]:
        """
        Validate Russian INN (Individual Tax Number)

        Args:
            inn: INN string to validate

        Returns:
            Tuple of (is_valid, error_message)
        """
        if not inn:
            return False, "INN is required"

        # Remove non-digits
        inn_clean = re.sub(r'\D', '', inn)

        # Check length (10 for organizations, 12 for individuals)
        if len(inn_clean) not in [10, 12]:
            return False, f"INN must be 10 digits (organizations) or 12 digits (individuals), got {len(inn_clean)}"

        # Validate checksum for 10-digit INN (organizations)
        if len(inn_clean) == 10:
            weights = [2, 4, 10, 3, 5, 9, 4, 6, 8]
            checksum = sum(int(inn_clean[i]) * weights[i] for i in range(9)) % 11
            if checksum > 9:
                checksum = checksum % 10

            if int(inn_clean[9]) != checksum:
                return False, "Invalid INN checksum for organization"

        # Validate checksum for 12-digit INN (individuals)
        elif len(inn_clean) == 12:
            # First checksum
            weights1 = [7, 2, 4, 10, 3, 5, 9, 4, 6, 8]
            checksum1 = sum(int(inn_clean[i]) * weights1[i] for i in range(10)) % 11
            if checksum1 > 9:
                checksum1 = checksum1 % 10

            if int(inn_clean[10]) != checksum1:
                return False, "Invalid INN checksum (first digit) for individual"

            # Second checksum
            weights2 = [3, 7, 2, 4, 10, 3, 5, 9, 4, 6, 8]
            checksum2 = sum(int(inn_clean[i]) * weights2[i] for i in range(11)) % 11
            if checksum2 > 9:
                checksum2 = checksum2 % 10

            if int(inn_clean[11]) != checksum2:
                return False, "Invalid INN checksum (second digit) for individual"

        return True, ""

    @staticmethod
    def validate_kpp(kpp: str) -> Tuple[bool, str]:
        """
        Validate Russian KPP (Tax Registration Reason Code)

        Args:
            kpp: KPP string to validate

        Returns:
            Tuple of (is_valid, error_message)
        """
        if not kpp:
            return False, "KPP is required for organizations"

        # Remove non-digits
        kpp_clean = re.sub(r'\D', '', kpp)

        # KPP must be exactly 9 digits
        if len(kpp_clean) != 9:
            return False, f"KPP must be exactly 9 digits, got {len(kpp_clean)}"

        # Basic format validation (first 4 digits - tax office code, next 2 - reason code, last 3 - sequential number)
        tax_office = kpp_clean[:4]
        reason_code = kpp_clean[4:6]

        # Tax office code should not be 0000
        if tax_office == "0000":
            return False, "Invalid tax office code in KPP"

        return True, ""

    @staticmethod
    def validate_ogrn(ogrn: str) -> Tuple[bool, str]:
        """
        Validate Russian OGRN (Primary State Registration Number)

        Args:
            ogrn: OGRN string to validate

        Returns:
            Tuple of (is_valid, error_message)
        """
        if not ogrn:
            return False, "OGRN is required"

        # Remove non-digits
        ogrn_clean = re.sub(r'\D', '', ogrn)

        # OGRN can be 13 digits (organizations) or 15 digits (individual entrepreneurs)
        if len(ogrn_clean) not in [13, 15]:
            return False, f"OGRN must be 13 digits (organizations) or 15 digits (entrepreneurs), got {len(ogrn_clean)}"

        # Validate checksum
        if len(ogrn_clean) == 13:
            # For organizations: check digit is remainder of (first 12 digits mod 11) mod 10
            first_12 = int(ogrn_clean[:12])
            check_digit = (first_12 % 11) % 10
            if int(ogrn_clean[12]) != check_digit:
                return False, "Invalid OGRN checksum for organization"

        elif len(ogrn_clean) == 15:
            # For entrepreneurs: check digit is remainder of (first 14 digits mod 13) mod 10
            first_14 = int(ogrn_clean[:14])
            check_digit = (first_14 % 13) % 10
            if int(ogrn_clean[14]) != check_digit:
                return False, "Invalid OGRN checksum for entrepreneur"

        return True, ""

    @staticmethod
    def validate_russian_postal_code(postal_code: str) -> Tuple[bool, str]:
        """
        Validate Russian postal code

        Args:
            postal_code: Postal code string to validate

        Returns:
            Tuple of (is_valid, error_message)
        """
        if not postal_code:
            return False, "Postal code is required"

        # Remove non-digits
        postal_clean = re.sub(r'\D', '', postal_code)

        # Russian postal codes are exactly 6 digits
        if len(postal_clean) != 6:
            return False, f"Russian postal code must be exactly 6 digits, got {len(postal_clean)}"

        # First digit should be 1-6 (Russian postal regions)
        if postal_clean[0] not in '123456':
            return False, "Invalid Russian postal code region"

        return True, ""


def validate_file_patterns(file_path: Path) -> List[str]:
    """
    Validate Russian business patterns in a Python file

    Args:
        file_path: Path to Python file to validate

    Returns:
        List of validation errors
    """
    errors = []

    try:
        content = file_path.read_text(encoding='utf-8')

        # Check for INN validation patterns
        inn_patterns = re.findall(r'inn["\']?\s*[:=]\s*["\']?([^"\'\\s,}]+)', content, re.IGNORECASE)
        for inn in inn_patterns:
            if inn and not inn.startswith('$') and not inn.startswith('{'):  # Skip template variables
                is_valid, error = RussianBusinessValidator.validate_inn(inn)
                if not is_valid:
                    errors.append(f"{file_path}: Invalid INN '{inn}' - {error}")

        # Check for KPP validation patterns
        kpp_patterns = re.findall(r'kpp["\']?\s*[:=]\s*["\']?([^"\'\\s,}]+)', content, re.IGNORECASE)
        for kpp in kpp_patterns:
            if kpp and not kpp.startswith('$') and not kpp.startswith('{'):
                is_valid, error = RussianBusinessValidator.validate_kpp(kpp)
                if not is_valid:
                    errors.append(f"{file_path}: Invalid KPP '{kpp}' - {error}")

        # Check for OGRN validation patterns
        ogrn_patterns = re.findall(r'ogrn["\']?\s*[:=]\s*["\']?([^"\'\\s,}]+)', content, re.IGNORECASE)
        for ogrn in ogrn_patterns:
            if ogrn and not ogrn.startswith('$') and not ogrn.startswith('{'):
                is_valid, error = RussianBusinessValidator.validate_ogrn(ogrn)
                if not is_valid:
                    errors.append(f"{file_path}: Invalid OGRN '{ogrn}' - {error}")

        # Check for postal code patterns
        postal_patterns = re.findall(r'postal_code["\']?\s*[:=]\s*["\']?([^"\'\\s,}]+)', content, re.IGNORECASE)
        for postal in postal_patterns:
            if postal and not postal.startswith('$') and not postal.startswith('{'):
                is_valid, error = RussianBusinessValidator.validate_russian_postal_code(postal)
                if not is_valid:
                    errors.append(f"{file_path}: Invalid postal code '{postal}' - {error}")

        # Check for VAT rate validation (should be reasonable for Russia)
        vat_patterns = re.findall(r'vat_rate["\']?\s*[:=]\s*["\']?([^"\'\\s,}]+)', content, re.IGNORECASE)
        for vat in vat_patterns:
            if vat and not vat.startswith('$') and not vat.startswith('{'):
                try:
                    vat_value = float(vat)
                    if vat_value < 0 or vat_value > 30:
                        errors.append(f"{file_path}: Unusual VAT rate '{vat}' - expected 0-30%")
                    elif vat_value not in [0, 10, 20]:  # Common Russian VAT rates
                        # This is a warning, not an error
                        pass
                except ValueError:
                    errors.append(f"{file_path}: Invalid VAT rate format '{vat}'")

    except Exception as e:
        errors.append(f"{file_path}: Error reading file - {str(e)}")

    return errors


def main():
    """Main validation function for command line usage"""
    if len(sys.argv) < 2:
        print("Usage: python russian_business.py <file1.py> [file2.py] ...")
        sys.exit(1)

    all_errors = []

    for file_arg in sys.argv[1:]:
        file_path = Path(file_arg)
        if file_path.exists() and file_path.suffix == '.py':
            errors = validate_file_patterns(file_path)
            all_errors.extend(errors)

    if all_errors:
        print("❌ Russian business validation errors found:")
        for error in all_errors:
            print(f"  {error}")
        sys.exit(1)
    else:
        print("✅ Russian business validation passed")
        sys.exit(0)


if __name__ == "__main__":
    main()

File: frontend/src/lib/validation/russian-business.ts
/**
 * Russian Business Validation Utilities
 * Validates INN, KPP, and OGRN numbers according to Russian business standards
 */

export interface ValidationResult {
  isValid: boolean;
  error?: string;
}

/**
 * Validates Russian INN (Individual Taxpayer Number)
 * Supports both 10-digit (organizations) and 12-digit (individuals) formats
 */
export function validateINN(inn: string): ValidationResult {
  if (!inn) {
    return { isValid: false, error: 'ИНН не может быть пустым' };
  }

  // Remove spaces and non-numeric characters
  const cleanINN = inn.replace(/\D/g, '');

  if (cleanINN.length === 10) {
    return validateINN10(cleanINN);
  } else if (cleanINN.length === 12) {
    return validateINN12(cleanINN);
  }

  return {
    isValid: false,
    error: 'ИНН должен содержать 10 цифр (для организаций) или 12 цифр (для ИП)',
  };
}

/**
 * Validates 10-digit INN for organizations
 */
function validateINN10(inn: string): ValidationResult {
  const weights = [2, 4, 10, 3, 5, 9, 4, 6, 8];

  let sum = 0;
  for (let i = 0; i < 9; i++) {
    sum += parseInt(inn[i]) * weights[i];
  }

  const checkDigit = sum % 11;
  const expectedCheckDigit = checkDigit < 10 ? checkDigit : 0;

  if (parseInt(inn[9]) !== expectedCheckDigit) {
    return { isValid: false, error: 'Неверная контрольная сумма ИНН' };
  }

  return { isValid: true };
}

/**
 * Validates 12-digit INN for individuals
 */
function validateINN12(inn: string): ValidationResult {
  // First check digit
  const weights1 = [7, 2, 4, 10, 3, 5, 9, 4, 6, 8];
  let sum1 = 0;
  for (let i = 0; i < 10; i++) {
    sum1 += parseInt(inn[i]) * weights1[i];
  }
  const checkDigit1 = sum1 % 11;
  const expectedCheckDigit1 = checkDigit1 < 10 ? checkDigit1 : 0;

  if (parseInt(inn[10]) !== expectedCheckDigit1) {
    return { isValid: false, error: 'Неверная первая контрольная сумма ИНН' };
  }

  // Second check digit
  const weights2 = [3, 7, 2, 4, 10, 3, 5, 9, 4, 6, 8];
  let sum2 = 0;
  for (let i = 0; i < 11; i++) {
    sum2 += parseInt(inn[i]) * weights2[i];
  }
  const checkDigit2 = sum2 % 11;
  const expectedCheckDigit2 = checkDigit2 < 10 ? checkDigit2 : 0;

  if (parseInt(inn[11]) !== expectedCheckDigit2) {
    return { isValid: false, error: 'Неверная вторая контрольная сумма ИНН' };
  }

  return { isValid: true };
}

/**
 * Validates Russian KPP (Tax Registration Reason Code)
 * 9-digit code for organizations only
 */
export function validateKPP(kpp: string): ValidationResult {
  if (!kpp) {
    return { isValid: false, error: 'КПП не может быть пустым' };
  }

  // Remove spaces and non-alphanumeric characters
  const cleanKPP = kpp.replace(/[^0-9A-Za-z]/g, '').toUpperCase();

  if (cleanKPP.length !== 9) {
    return { isValid: false, error: 'КПП должен содержать 9 символов' };
  }

  // First 4 characters should be digits (tax office code)
  const taxOfficeCode = cleanKPP.substring(0, 4);
  if (!/^\d{4}$/.test(taxOfficeCode)) {
    return { isValid: false, error: 'Первые 4 символа КПП должны быть цифрами' };
  }

  // Next 2 characters should be alphanumeric (reason code)
  const reasonCode = cleanKPP.substring(4, 6);
  if (!/^[0-9A-Z]{2}$/.test(reasonCode)) {
    return { isValid: false, error: 'Символы 5-6 КПП должны быть цифрами или буквами' };
  }

  // Last 3 characters should be digits (record number)
  const recordNumber = cleanKPP.substring(6, 9);
  if (!/^\d{3}$/.test(recordNumber)) {
    return { isValid: false, error: 'Последние 3 символа КПП должны быть цифрами' };
  }

  return { isValid: true };
}

/**
 * Validates Russian OGRN (Primary State Registration Number)
 * 13 digits for organizations, 15 digits for entrepreneurs (OGRNIP)
 */
export function validateOGRN(ogrn: string): ValidationResult {
  if (!ogrn) {
    return { isValid: false, error: 'ОГРН не может быть пустым' };
  }

  // Remove spaces and non-numeric characters
  const cleanOGRN = ogrn.replace(/\D/g, '');

  if (cleanOGRN.length === 13) {
    return validateOGRN13(cleanOGRN);
  } else if (cleanOGRN.length === 15) {
    return validateOGRNIP15(cleanOGRN);
  }

  return {
    isValid: false,
    error: 'ОГРН должен содержать 13 цифр (для организаций) или 15 цифр (для ИП)',
  };
}

/**
 * Validates 13-digit OGRN for organizations
 */
function validateOGRN13(ogrn: string): ValidationResult {
  const firstTwelveDigits = ogrn.substring(0, 12);
  const checkDigit = parseInt(ogrn[12]);

  const remainder = BigInt(firstTwelveDigits) % BigInt(11);
  const expectedCheckDigit = Number(remainder % BigInt(10));

  if (checkDigit !== expectedCheckDigit) {
    return { isValid: false, error: 'Неверная контрольная сумма ОГРН' };
  }

  return { isValid: true };
}

/**
 * Validates 15-digit OGRNIP for entrepreneurs
 */
function validateOGRNIP15(ogrnip: string): ValidationResult {
  const firstFourteenDigits = ogrnip.substring(0, 14);
  const checkDigit = parseInt(ogrnip[14]);

  const remainder = BigInt(firstFourteenDigits) % BigInt(13);
  const expectedCheckDigit = Number(remainder % BigInt(10));

  if (checkDigit !== expectedCheckDigit) {
    return { isValid: false, error: 'Неверная контрольная сумма ОГРНИП' };
  }

  return { isValid: true };
}

/**
 * Validates complete Russian business entity data
 */
export interface BusinessEntityData {
  inn: string;
  kpp?: string; // Optional for individuals
  ogrn: string;
  companyName: string;
}

export function validateBusinessEntity(data: BusinessEntityData): ValidationResult {
  // Validate INN
  const innResult = validateINN(data.inn);
  if (!innResult.isValid) {
    return innResult;
  }

  // Validate OGRN
  const ogrnResult = validateOGRN(data.ogrn);
  if (!ogrnResult.isValid) {
    return ogrnResult;
  }

  // For organizations (10-digit INN), KPP is required
  const cleanINN = data.inn.replace(/\D/g, '');
  if (cleanINN.length === 10) {
    if (!data.kpp) {
      return { isValid: false, error: 'КПП обязателен для организаций' };
    }

    const kppResult = validateKPP(data.kpp);
    if (!kppResult.isValid) {
      return kppResult;
    }
  }

  // Validate company name
  if (!data.companyName || data.companyName.trim().length < 2) {
    return { isValid: false, error: 'Название организации должно содержать минимум 2 символа' };
  }

  return { isValid: true };
}

/**
 * Formats INN with appropriate spacing
 */
export function formatINN(inn: string): string {
  const cleanINN = inn.replace(/\D/g, '');

  if (cleanINN.length === 10) {
    // Format as XXXX XXXXXX for organizations
    return cleanINN.replace(/(\d{4})(\d{6})/, '$1 $2');
  } else if (cleanINN.length === 12) {
    // Format as XXXX XXXX XXXX for individuals
    return cleanINN.replace(/(\d{4})(\d{4})(\d{4})/, '$1 $2 $3');
  }

  return cleanINN;
}

/**
 * Formats KPP with appropriate spacing
 */
export function formatKPP(kpp: string): string {
  const cleanKPP = kpp.replace(/[^0-9A-Za-z]/g, '').toUpperCase();

  if (cleanKPP.length === 9) {
    // Format as XXXX XX XXX
    return cleanKPP.replace(/(\d{4})([0-9A-Z]{2})(\d{3})/, '$1 $2 $3');
  }

  return cleanKPP;
}

/**
 * Formats OGRN with appropriate spacing
 */
export function formatOGRN(ogrn: string): string {
  const cleanOGRN = ogrn.replace(/\D/g, '');

  if (cleanOGRN.length === 13) {
    // Format as X XX XX XXXXXXX for organizations
    return cleanOGRN.replace(/(\d{1})(\d{2})(\d{2})(\d{7})(\d{1})/, '$1 $2 $3 $4 $5');
  } else if (cleanOGRN.length === 15) {
    // Format as X XX XX XXXXXXXXX for entrepreneurs
    return cleanOGRN.replace(/(\d{1})(\d{2})(\d{2})(\d{9})(\d{1})/, '$1 $2 $3 $4 $5');
  }

  return cleanOGRN;
}

/**
 * Determines if INN belongs to an organization or individual
 */
export function getINNType(inn: string): 'organization' | 'individual' | 'invalid' {
  const cleanINN = inn.replace(/\D/g, '');

  if (cleanINN.length === 10) {
    return 'organization';
  } else if (cleanINN.length === 12) {
    return 'individual';
  }

  return 'invalid';
}

/**
 * VAT validation for Russian business
 */
export function validateVATRate(vatRate: number): ValidationResult {
  const validRates = [0, 10, 20]; // Valid VAT rates in Russia

  if (!validRates.includes(vatRate)) {
    return {
      isValid: false,
      error: 'Недопустимая ставка НДС. Разрешенные значения: 0%, 10%, 20%',
    };
  }

  return { isValid: true };
}

/**
 * Currency validation for Russian business
 */
export function validateCurrency(currency: string): ValidationResult {
  const validCurrencies = ['RUB', 'CNY', 'USD', 'EUR']; // Supported currencies

  if (!validCurrencies.includes(currency.toUpperCase())) {
    return {
      isValid: false,
      error: 'Недопустимая валюта. Поддерживаемые валюты: RUB, CNY, USD, EUR',
    };
  }

  return { isValid: true };
}
